[RFC] Implement caching for user defined triton kernels

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140326

This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

oulgen committed:
Implement caching for user defined triton kernels

[ghstack-poisoned]

oulgen committed:
Update on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1544,75 +1622,9 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
-        # global constexpr vars handled above
-        symbols_included = {original_name}
-
-        def traverse(cur_kernel):
```
any chance of doing this code move as a refactor pr stacked before this? Hard to review like this

oulgen replied:
done

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -757,6 +757,20 @@ def trace_triton_kernel_wrapper(
+    )
+
+    kernel_source = user_defined_triton_kernel_transitive_closure_source_code(kernel)
+    constant_args = kernel_side_table.get_constant_args(proxy_args["constant_args_idx"])
```
We should change how the constant args works (to also benefit the triton kernel export work). Constant args should be inputs to the node, unless they are non-fx-graphable types. The non-fx-graphable types should go into the side table.

oulgen replied:
I guess I didnt want constant args to be spit between on graph and on side table, wanted to keep them all together originally

zou3519 replied:
We already split the constant args to do triton kernel non-strict export, because non-strict export doesn't have the concept of "constant args"

zou3519 commented on torch/_inductor/codecache.py:
```diff
@@ -615,6 +619,25 @@ def _reduce_unsupported(self, s: Any) -> NoReturn:
+        """
+        fn, (data, imports) = gm.__reduce__()
+        code = data["_code"]
+        code = re.sub(r"kernel_idx = \d+", "", code)
+        code = re.sub(r"constant_args_idx = \d+", "", code)
```
maybe these should just both be hashes so we don't need to do the regex delete thing

zou3519 replied:
how does fx_graph_cache handle graphs that look similar but have different variable names? I think it's reasonable to have some sort of "graph normalization pass"

oulgen replied:
it does not, unless the graph are identical, we cache miss. I thought this was the safest starting point. We could do some normalization eventually but i want to understand how much does this bring us

zou3519 commented on torch/_inductor/codecache.py:
```diff
@@ -776,6 +799,32 @@ def __init__(
+                        node.target,
+                        (TritonKernelWrapperFunctional, TritonKernelWrapperMutation),
+                    ):
+                        data = node.meta.get(
+                            "user_defined_triton_kernel_source_and_constant_args", None
+                        )
```
@oulgen We've got some options:
- toss the triton kernel source into node.meta
- toss the triton kernel source into graph.meta
- toss the triton kernel source as a string input to the node (we don't need to read from the source if the triton kernel is still alive; we can also probably develop some mechanism for this to print in a sane way).

Ditto for the constant args (in option 3, if the constant args are not fx-graphable, then we could just say they're not cacheable).

One question is: how much do you care about the triton kernel export serializability work? There could be one solution that works for both triton kernel export serializability and the caching work.

zou3519 replied:
I kind of like storing the triton kernel source in the graph.meta -- the claim here is that we are essentially going to move the side table to be in the graph itself.

oulgen replied:
I think what you're suggesting is a big refactor to get rid of side table all together which likely requires us to eval the source code for inductor so that we have a function pointer again, or store all the kernel properties ahead of time. For caching purposes only, i wanted to start with just the source since we dont need to restore anything but only use it as a key.

The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

zou3519 replied:
My suggestion is more: I'm worried that we're going to end up with two different implementations, one for caching and one for serialization, or that when we implement serialization we will end up re-implementing caching. Since the timeline for serialization is also soon (@ydwu4's planning on working on it in the next couple of weeks?) I wonder if it makes sense to slow down and figure out how to co-design these. Feel free to make an argument that the code in this PR is easy to rip out, for example.

I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

> The problem with storing in graph.meta is if someone dropped the meta or removed one in an opt pass, we can no longer assert that we have all of them, as opposed to the current assertion. wdyt?

In either (1) or (2) it seems like we need to build some assertions into the system. For node.meta we would need to build assertions that for all triton_kernel_* nodes, the node.meta must be populated. For graph.meta, we would need to build the assertion that if the graph has any triton_kernel_* nodes, the triton kernels they are referring to have their sources in graph.meta. A benefit of building into graph.meta is that we can deduplicate nodes (idk if that actually matters for things like cache key computation)

aakhundov replied:
@zou3519 I agree eval is undesirable (e.g., as we'd need to collect all transitive dependencies, relevant imports etc.; I think, we rely on `Autotuner` and `JITFunction` instances in multiple places, so we'd need to be able to reliably recreate them from the source).

> I'm not suggesting that we should eval the source code for inductor -- I can see how that would be undesirable.

Given just the source, how can we then reconstruct the kernel instance w/o eval?

aakhundov replied:
Did we consider serializing the side table instead of the kernel source? Serialized Triton kernel source (in Python) makes little sense w/o Python runtime, and with it we could also deserialize the kernel instances in the side table?

oulgen replied:
@zou3519 yep your points make sense, but I do worry that this requires a larger refactor around user defined triton kernels, and i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph? Moving the assertions to graph instead of node is trickier which is why i chose this for the quick-and-dirty solution. If you think this is too much tech debt, I'm happy to accept that.

oulgen replied:
@aakhundov Deserialization is the hard question here, esp if we want to do that without eval, by storing a ton of metadata, and since I didnt need it for hashing, I tried to avoid it. But I dont think serde of the side table is any harder than serde of a single kernel pointer, so i think it boils down to the same problem.

zou3519 replied:
>  i'm primarily trying to unblock ads models as their compile time reduction is blocked on lack of user defined triton kernel caching. Would you be willing to proceed with this and later rip this out or move the assertions to graph?

Yes I'm happy to proceed with the current design, esp bc there is a pressing need for this right now

oulgen committed:
Update on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

oulgen commented on torch/_inductor/codecache.py:
```diff
@@ -614,6 +619,25 @@ def _reduce_unsupported(self, s: Any) -> NoReturn:
+        fn, (data, imports) = gm.__reduce__()
+        code = data["_code"]
+        code = re.sub(r"kernel_idx = \d+", "", code)
+        code = re.sub(r"constant_args_idx = \d+", "", code)
```
idx is just an entry into the table, we graph that entry and convert it to source code. we add the transitive closure of source code of each triton kernel in this graph, we assert that we do it for each triton op on the graph.

ezyang replied:
Err, isn't the real reason because you shoved it in `user_defined_triton_kernel_source_and_constant_args`

oulgen replied:
Yea, that's what I said, no? grab the entry put it into node.meta and later put it in the cache key

zou3519 replied:
I think you guys are saying the same thing lol

zou3519 commented on torch/_functorch/_aot_autograd/autograd_cache.py:
```diff
@@ -275,7 +275,7 @@ def autograd_cache_key(
     check_cacheable(gm)
     details = AOTAutogradCacheDetails(gm, example_inputs, config, fx_config)
-    pickler = AOTAutogradCachePickler()
+    pickler = AOTAutogradCachePickler(gm)
```
Are you enabling this for AOTAutogradCache too, or is that another PR? 

If so, Dynamo produces triton_kernel_wrapper_mutation or a triton_op and I don't see those being handling here

oulgen replied:
This PR is just FXGraphCache, I needed to update the pickling so this is just conforming the signature. I can do AOTAutograd as a follow up

zou3519 commented on test/inductor/test_codecache.py:
```diff
@@ -494,13 +494,41 @@ def fn2(q, k, v):
+    @config.patch({"fx_graph_cache": True})
+    @config.patch({"fx_graph_remote_cache": False})
+    @parametrize("bundle_triton", (False, True))
+    def test_triton_higher_order_op(self, bundle_triton):
```
For testing:
- can we test that calling a function with a different triton kernel results in a cache miss?

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -638,7 +638,7 @@ def __call__(
 class TritonKernelWrapperFunctional(HigherOrderOperator):
     def __init__(self) -> None:
-        super().__init__("triton_kernel_wrapper_functional", cacheable=False)
+        super().__init__("triton_kernel_wrapper_functional", cacheable=True)
```
I'm confused, FXGraphCache should only see one of {triton_kernel_wrapper_mutation, triton_kernel_wrapper_functional} (depending on where the cache is in Inductor), did we need to make both cacheable?

zou3519 replied:
No harm, the problem is mostly that the test coverage only covers one of them so if we didn't actually implement the other one correctly and it somehow shows up in in the future then problems

zou3519 reviewed (commented):
Given our discussion, the approach LGTM. I want to check if we're changing the behavior for AOTAutogradCache since there are some code changes there, also added some testing suggestions

ezyang commented:
Deferring to Richard for approval

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen commented:
@zou3519 added unit test for cache miss, will do aotautograd as a follow up

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).

oulgen commented:
@pytorchbot merge

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen committed:
None on "[RFC] Implement caching for user defined triton kernels"


This PR adds caching for user defined triton kernels by putting the transitive closure of source code in node.meta along with constant arguments.

One HUGE hack we do here is a node looks like
```
triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(1, 1, 1)], tma_descriptor_
metadata = {}, kwargs = {'in_ptr0': arg0_1, 'in_ptr1': arg1_1, 'out_ptr': arg0_1}, tensors_to_clone = ['out_ptr']);
```
so we use regex to remove `kernel_idx = 0, constant_args_idx = 1` parts as they are not relevant to cache hash. This is horrible and I'd like to eventually not use pickle as a hashing alternative but this is a longer project.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

Differential Revision: [D65895744](https://our.internmc.facebook.com/intern/diff/D65895744)

[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65895744).
